### PR TITLE
Feature/sg 216 refactor imagenet dataloader

### PR DESCRIPTION
**Pull request by Louis-Dupont**

### CODE REVIEW

File: src/super_gradients/recipes/checkpoint_params/default_checkpoint_params.yaml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 load_checkpoint: False # whether to load checkpoint
 load_backbone: False # whether to load only backbone part of checkpoint
+checkpoint_path: # checkpoint path that is located in super_gradients/checkpoints
```
Comment (shaydeci) [inline src/super_gradients/recipes/checkpoint_params/default_checkpoint_params.yaml:3]: Whats this doing here ?

File: src/super_gradients/recipes/dataset_params/imagenet_mobilenetv2_dataset_params.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+train_loader_drop_last: True
+batch_size: 256
+val_batch_size: 256
+random_erase_prob: 0.2
+random_erase_value: random
+train_interpolation: random
+rand_augment_config_string: rand-m9-mstd0.5
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_mobilenetv2_dataset_params.yaml:12]: Why is this commented out ?

File: src/super_gradients/recipes/dataset_params/imagenet_mobilenetv3_dataset_params.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+defaults:
+  - imagenet_dataset_params
+
+batch_size: 128
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
+#defaults:
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_mobilenetv3_dataset_params.yaml:7]: Redundant commented out code.

File: src/super_gradients/recipes/dataset_params/imagenet_vit_base_dataset_params.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  mixup_alpha: 0.2
+  cutmix_alpha: 1.0
+  label_smoothing: 0.1
+img_mean: [0.5, 0.5, 0.5]
+img_std: [0.5, 0.5, 0.5]
+
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
+
+# RuntimeError: Default process group has not been initialized, please make sure to call init_process_group.
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_vit_base_dataset_params.yaml:21]: What is this ?

File: src/super_gradients/recipes/imagenet_vit_large.yaml [hunk 5]
```diff
@@ -15,9 +15,9 @@
 defaults:
   - imagenet_vit_base
 
-
 dataset_params:
-  batch_size: 32
+  train_dataloader_params:
```
Comment (shaydeci) [inline src/super_gradients/recipes/imagenet_vit_large.yaml:19]: This doesnt change the dataset_params which are actually used now in train_from_recipe. You are changing the recipe by doing this !

File: src/super_gradients/training/datasets/classification_datasets/image_net.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import torchvision.datasets as torch_datasets
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/classification_datasets/image_net.py:1]: Please rename the file to imagenet_dataset.py.

File: src/super_gradients/common/factories/transforms_factory.py [hunk 7]
```diff
@@ -31,8 +37,12 @@ def __init__(self):
             "DetectionHorizontalFlip": DetectionHorizontalFlip,
             "DetectionPaddedRescale": DetectionPaddedRescale,
             "DetectionTargetsFormat": DetectionTargetsFormat,
-            "DetectionTargetsFormatTransform": DetectionTargetsFormatTransform
+            "DetectionTargetsFormatTransform": DetectionTargetsFormatTransform,
 
+            'RandomResizedCropAndInterpolation': RandomResizedCropAndInterpolation,
+            'color_augmentation': datasets_utils.get_color_augmentation,
```
Comment (shaydeci) [inline src/super_gradients/common/factories/transforms_factory.py:43]: This is just a name we used previously (and a pretty bad one tbh) to get either RandAugment or color jitter.
We should get rid of this and set rand augment (also addd it to the factory) when we use it and color jitter separately.

File: src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml [hunk 8]
```diff
@@ -12,4 +12,86 @@ imagenet_pca_aug: 0.0 # imagenet pca augmentation (applied only to trainset)
 train_interpolation: default # interpolation mode
 rand_augment_config_string: # randaugment config string (see super_gradients/training/datasets/auto_augment.py)
 random_erase_prob: 0.0 # random erase probability (applied only to trainset)
-aug_repeat_count: 0 # amount of repetitions (each repetition of an example is augmented differently) for a trainset example.
\ No newline at end of file
+aug_repeat_count: 0 # amount of repetitions (each repetition of an example is augmented differently) for a trainset example.
+
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
+#dataset_dir: /data/Imagenet     # path to imagenet directory (local)
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml:19]: What are all these commented out lines ?

File: src/super_gradients/recipes/dataset_params/imagenet_resnet50_dataset_params.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  mixup_alpha: 0.2
+  cutmix_alpha: 1.0
+  label_smoothing: 0.1
+
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
+#defaults:
+#  - imagenet_dataset_params
+
+train_transform_args:
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_resnet50_dataset_params.yaml:20]: This is definitely not what we are going for...
We wanted the transforms to be visible for the user and this does exactly the same as dataset interface.

Its ok that we have some parenting args to help the user easily adapt transforms (such as input size), but definitely not this.

File: src/super_gradients/recipes/dataset_params/imagenet_resnet50_kd_dataset_params.yaml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+cutmix: True
+cutmix_params:
+  mixup_alpha: 0.2
+  cutmix_alpha: 1.0
+  label_smoothing: 0.1
+aug_repeat_count: 3
+
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
+# RuntimeError: Default process group has not been initialized, please make sure to call init_process_group.
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_resnet50_kd_dataset_params.yaml:19]: What ?

**Review by shaydeci — CHANGES_REQUESTED** [review]
Inline.

File: src/super_gradients/recipes/checkpoint_params/default_checkpoint_params.yaml (hunk 1 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/checkpoint_params/default_checkpoint_params.yaml:3]: It seems that it is now required to have "checkpoint_path" defined (even if it's value is None):
Check https://github.com/Deci-AI/super-gradients/blob/fe011b2172768a17a9795071647c5efe6041da4f/src/super_gradients/training/sg_trainer/sg_trainer.py#L210-L217

File: src/super_gradients/recipes/dataset_params/imagenet_mobilenetv3_dataset_params.yaml (hunk 3 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_mobilenetv3_dataset_params.yaml:8]: When we will remove the section above we need to keep the defaults. In other sections we have other params that we also want to keep (like img_mean).
To make sure that we don't forget these params when we will remove the section above I already wrote and commented the lines of code that we will still need.
I added a small comment to explain that this code should be uncommented it when the above part is removed: dd7a66c25443ec5d09d590c2c92cc6750e0f5cd7

File: src/super_gradients/recipes/dataset_params/imagenet_vit_base_dataset_params.yaml (hunk 4 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_vit_base_dataset_params.yaml]: This was a note while debugging, I removed it

File: src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml (hunk 8 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml]: Same as comment above

File: src/super_gradients/recipes/imagenet_vit_large.yaml (hunk 5 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/imagenet_vit_large.yaml:20]: You are right, i added batch_size again in b55399ddcb5c2994ef47ee02d68414563170817e

File: src/super_gradients/recipes/dataset_params/imagenet_resnet50_kd_dataset_params.yaml (hunk 10 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_resnet50_kd_dataset_params.yaml]: This was for debugging, I removed it

File: src/super_gradients/recipes/dataset_params/imagenet_resnet50_dataset_params.yaml (hunk 9 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_resnet50_dataset_params.yaml:21]: We talked about it and the problem is that we cannot modify an element in a list, so if we want to modify any transform we have to redefine the whole list of transforms.

So either we always repeat all the transforms (redundant but explicit)
Or we create this "train_transform_args", that acts like an interface to not have to repeat all the transforms (shorter but implicit)

File: src/super_gradients/training/datasets/classification_datasets/image_net.py (hunk 6 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/datasets/classification_datasets/image_net.py]: changed in 1cbca1572e70847368012225c96711905ac2fd9f

File: src/super_gradients/recipes/dataset_params/imagenet_mobilenetv2_dataset_params.yaml (hunk 2 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_mobilenetv2_dataset_params.yaml:12]: see other comment

File: src/super_gradients/recipes/dataset_params/imagenet_resnet50_dataset_params.yaml (hunk 9 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_resnet50_dataset_params.yaml]: I changed according to our discussion: redefine all the transforms when one is changed, in order to have explicit and easy to understand recipes

File: src/super_gradients/common/factories/transforms_factory.py (hunk 7 above)
Comment (Louis-Dupont) [reply src/super_gradients/common/factories/transforms_factory.py]: Changed accordingly

File: src/super_gradients/common/factories/transforms_factory.py [hunk 11]
```diff
@@ -31,8 +36,12 @@ def __init__(self):
             "DetectionHorizontalFlip": DetectionHorizontalFlip,
             "DetectionPaddedRescale": DetectionPaddedRescale,
             "DetectionTargetsFormat": DetectionTargetsFormat,
-            "DetectionTargetsFormatTransform": DetectionTargetsFormatTransform
+            "DetectionTargetsFormatTransform": DetectionTargetsFormatTransform,
 
+            'RandomResizedCropAndInterpolation': RandomResizedCropAndInterpolation,
+            'RandomAugmentation': rand_augment_transform,
```
Comment (shaydeci) [inline src/super_gradients/common/factories/transforms_factory.py:42]: This isn't just "random augmentation", the name RandAugment has a meaning in literature.
Please rename it to RandAugmentTransform.

File: src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml [hunk 12]
```diff
@@ -12,4 +12,53 @@ imagenet_pca_aug: 0.0 # imagenet pca augmentation (applied only to trainset)
...
+
+# TODO: REMOVE ABOVE, HERE FOR COMPATIBILITY UNTIL WE REMOVE DATASET_INTERFACE
+# TODO: UNCOMMENT BELOW WHEN ABOVE IS REMOVED
+# Base recipe for ImageNet Datasets amd Dataloaders.
+#img_mean: [0.485, 0.456, 0.406] # mean for normalization
+#img_std: [0.229, 0.224, 0.225]  # std  for normalization
+
+train_dataset_params:
+  root: /data/Imagenet/train
+  transform:
```
Comment (shaydeci) [inline src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml:26]: Lets align imagenet dataset with the rest by just passing transforms (i.e a list).
We can then wrap it up with Compost in ImagenetDataset's __init__

**Review by shaydeci — CHANGES_REQUESTED** [review]
2 small comments.

File: src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml (hunk 12 above)
Comment (Louis-Dupont) [reply src/super_gradients/recipes/dataset_params/imagenet_dataset_params.yaml]: Changed in 007ebe69dcfae3d8ded7874f243b7cbaab5c4b25

**Review by shaydeci — APPROVED** [review]
LGTM

**Pull request merged**